File: DATA_STRUCTURES_CPP/Heap___(MAX_and_MIN).cpp
/* MAX HEAP - Максимальная куча (самый большой ключ находится в корне) */
// Ключи родительских узлов всегда больше или равны ключам дочерних узлов
/* MIN HEAP - Минимальная куча (самый меньший ключ находится в корне) */
// Ключи родительских узлов меньше или равны ключам дочерних узлов
#include <iostream>
#include <stdexcept>
#include <vector>

using std::vector;
using std::out_of_range;
using std::swap;
using std::cout;
using std::endl;


template <typename Type>
class Heap_max {
protected:
    vector<Type> heap_elem;                                   // vector для хранения элементов кучи
    int get_parent_id(const int& id) { return (id - 1) / 2;}  // Возвращает индекс родителя
    int get_rchild_id(const int& id) { return (2 * id + 2);}  // Возвращает индекс правого потомка
    int get_lchild_id(const int& id) { return (2 * id + 1);}  // Возвращает индекс левого потомка
    void heapify_down(const int& id);                         // Рекурсивно перемещает элемент Array[id] вниз по дереву
    void heapify_up(const int& id);                           // Перемещает Array[id] вверх по дереву
public:
    const size_t size() const { return heap_elem.size(); }    // Возвращаем размер кучи
    bool empty() const { return heap_elem.size() == 0; }      // true если пуста, иначе false
    void push(const Type &key);                               // Добавить элемент в кучу
    void pop();                                               // Удалить элемент из кучи с наивысшим приоритетом
    Type top();                                               // Получить элемент с наивысшим приоритетом
};

template <typename Type>
class Heap_min {
private:
    vector<Type> heap_elem;                                   // vector для хранения элементов кучи
    int get_parent_id(const int& id) { return (id - 1) / 2;}  // Возвращает индекс родителя
    int get_rchild_id(const int& id) { return (2 * id + 2);}  // Возвращает индекс правого потомка
    int get_lchild_id(const int& id) { return (2 * id + 1);}  // Возвращает индекс левого потомка
    void heapify_down(const int& id);                         // Рекурсивно перемещает элемент Array[id] вниз по дереву
    void heapify_up(const int& id);                           // Перемещает Array[id] вверх по дереву
public:
    const size_t size() const { return heap_elem.size(); }    // Возвращаем размер кучи
    bool empty() const { return heap_elem.size() == 0; }      // true если пуста, иначе false
    void push(const Type& key);                               // Добавить элемент в кучу
    void pop();                                               // Удалить элемент из кучи с наивысшим приоритетом
    Type top();                                               // Получить элемент с наивысшим приоритетом
};

/************************MAX HEAP************************/
template <typename Type>
void Heap_max<Type>::heapify_down(int const &id) {
    int left = get_lchild_id(id);          // Получить левого потомка узла с индексом 'id'
    int right = get_rchild_id(id);        // Получить правого потомка узла с индексом 'id'
    int largest = id;                       // Запомнить 'id' наибольшего узла

    // Cравниваем Array[id] с его левым и правым дочерними элементами и находим наибольшее значение
    if (left < size() && heap_elem[left] > heap_elem[id])
        largest = left;
    if (right < size() && heap_elem[right] > heap_elem[largest])
        largest = right;
    if (largest != id) {                         // Если индекс наибольшего и найденого не равны, иначе ничего не делать
        swap(heap_elem[id], heap_elem[largest]); // Поменять местами потомка и родителя
        heapify_down(largest);                   // И вызвать рекурсивно heapify-down для потомка
    }
}

template <typename Type>
void Heap_max<Type>::heapify_up(const int &id) {
    if (id && heap_elem[get_parent_id(id)] < heap_elem[id]) {  // Ели узел с индексом 'id' и его родитель нарушают свойство кучи
        swap(heap_elem[id], heap_elem[get_parent_id(id)]);     // Поменять их местами, если свойство кучи нарушено
        heapify_up(get_parent_id(id));                         // Вызвать heapify-up для родителя
    }
}

template <typename Type>
void Heap_max<Type>::push(const Type &key) {
    heap_elem.push_back(key);                   // Вставить новый элемент в конец вектора
    int index = static_cast<int>(size() - 1);   // Получить индекс элемента
    heapify_up(index);                          // Вызов heapify-up от index
}

template <typename Type>
void Heap_max<Type>::pop() {
    try {
        if (size() == 0) {
            throw out_of_range("Heap is emty!");    // Если в куче нет элементов, бросить исключение, иначе
        }
        heap_elem[0] = heap_elem.back();            // Заменить корень кучи последним элементом вектора
        heap_elem.pop_back();                       // И удалить заменённый элемент
        heapify_down(0);                            // Вызвать heapify-down для корня
    }
    catch (const out_of_range& msg) {               // Ловим и печатаем исключение
        cout << endl << msg.what();
    }
}

template <typename Type>
Type Heap_max<Type>::top() {
    try {
        if (size() == 0) {              // Если в куче нет элементов, бросить исключение,
            throw out_of_range("Heap is empty, stupid man!");
        }
        return heap_elem.at(0);         // Иначе возвращаем верхний (первый) элемент
    }
    catch (const out_of_range& msg) {   // ловим и печатаем исключение
        cout << endl << msg.what();
        return Type(0);
    }
}

/************************MIN HEAP************************/
template <typename Type>
void Heap_min<Type>::heapify_down(int const& id) {
    int left = get_lchild_id(id);         // Получить левого потомка узла с индексом 'id'
    int right = get_rchild_id(id);        // Получить правого потомка узла с индексом 'id'
    int smallest = id;                    // Запомнить 'id' НАИМЕНЬШЕГО узла

    // Cравниваем Array[id] с его левым и правым дочерними элементами и находим наибольшее значение
    if (left < size() && heap_elem[left] < heap_elem[id])
        smallest = left;
    if (right < size() && heap_elem[right] < heap_elem[smallest])
        smallest = right;
    if (smallest != id) {                         // Если индекс наибольшего и найденого не равны, иначе ничего не делать
        swap(heap_elem[id], heap_elem[smallest]); // Поменять местами потомка и родителя
        heapify_down(smallest);                   // И вызвать рекурсивно heapify-down для потомка
    }
}

template <typename Type>
void Heap_min<Type>::heapify_up(const int& id) {
    if (id && heap_elem[get_parent_id(id)] > heap_elem[id]) {  // Ели узел с индексом 'id' и его родитель нарушают свойство кучи
        swap(heap_elem[id], heap_elem[get_parent_id(id)]);     // Поменять их местами, если свойство кучи нарушено
        heapify_up(get_parent_id(id));                         // Вызвать heapify-up для родителя
    }
}

template <typename Type>
void Heap_min<Type>::push(const Type& key) {
    heap_elem.push_back(key);                   // Вставить новый элемент в конец вектора
    int index = static_cast<int>(size() - 1);   // Получить индекс элемента
    heapify_up(index);                          // Вызов heapify-up от index
}

template <typename Type>
void Heap_min<Type>::pop() {
    try {
        if (size() == 0) {
            throw out_of_range("Heap is emty!");    // Если в куче нет элементов, бросить исключение, иначе
        }
        heap_elem[0] = heap_elem.back();            // Заменить корень кучи последним элементом вектора
        heap_elem.pop_back();                       // И удалить заменённый элемент
        heapify_down(0);                            // Вызвать heapify-down для корня
    }
    catch (const out_of_range& msg) {               // Ловим и печатаем исключение
        cout << endl << msg.what();
    }
}

template <typename Type>
Type Heap_min<Type>::top() {
    try {
        if (size() == 0) {              // Если в куче нет элементов, бросить исключение,
            throw out_of_range("Heap is empty, stupid man!");
        }
        return heap_elem.at(0);         // Иначе возвращаем верхний (первый) элемент
    }
    catch (const out_of_range& msg) {   // ловим и печатаем исключение
        cout << endl << msg.what();
        return Type(0);
    }
}


int main() {
    setlocale(0, "");
    Heap_max<int> iheap;    Heap_max<double> dheap;
    Heap_min<int> imin;     Heap_min<double> dmin;

    for (int i = 0; i < 20; ++i) {
        iheap.push(rand() % 100);   dheap.push(i * 0.28);
        imin.push(rand() % 100);    dmin.push(i * 0.28);
    }
    cout << "Int max heap: ";
    for (int i = 0; i < 20; ++i) {
        cout << iheap.top() << " ";
        iheap.pop();
    }
    cout << "\nInt min heap: ";
    for (int i = 0; i < 20; ++i) {
        cout << imin.top() << " ";
        imin.pop();
    }
    cout << "\nDouble max heap: ";
    for (int i = 0; i < 20; ++i) {
        cout << dheap.top() << " ";
        dheap.pop();
    }
    cout << "\nDouble min heap: ";
    for (int i = 0; i < 20; ++i) {
        cout << dmin.top() << " ";
        dmin.pop();
    }
    cout << "\nChecking for emptiness:";
    dheap.top();
    iheap.top();
    return 0;
}

File: DATA_STRUCTURES_PY/Heap___(MAX_and_MIN).py
# MAX HEAP - Максимальная куча (самый большой ключ находится в корне)
# Ключи родительских узлов всегда больше или равны ключам дочерних узлов
# MIN HEAP - Минимальная куча (самый маленький ключ находится в корне)
# Ключи родительских узлов меньше или равны ключам дочерних узлов
import random

class Heap_min: pass                        # Класс описывающий минимальную кучу (корень самый маленький элемент)
class Heap_max:                             # Класс описывающий кучу (корень самый большой элемент)
    def __init__(self): pass                # Конструктор
    def __get_parent_id(self, ind): pass    # Получить индекс родителя
    def __get_rchild_id(self, ind): pass    # Получить индекс правого потомка
    def __get_lchild_id(self, ind): pass    # Получить индекс левого потомка
    def size(self): pass                    # Получить размер кучи
    def _heapify_down(self, ind): pass      # Проталкивание элемента вниз
    def _heapify_up(self, ind): pass        # Проталкивание элемента вверх
    def push(self, key): pass               # Добавление элемента
    def pop(self): pass                     # Удаление
    def top(self): pass                     # Получить верхний элемент (корень)
"""-----------------------------------------------------------"""


class Heap_max:
    def __init__(self):
        self._heap_elem = []

    def __get_parent_id(self, ind):
        return (ind - 1) // 2

    def __get_rchild_id(self, ind):
        return 2 * ind + 2

    def __get_lchild_id(self, ind):
        return 2 * ind + 1

    def size(self):
        return len(self._heap_elem)

    def _heapify_down(self, ind):
        left = self.__get_lchild_id(ind)    # Получить левого потомка узла с индексом 'id'
        right = self.__get_rchild_id(ind)   # Получить правого потомка узла с индексом 'id'
        largest = ind                       # Запомнить 'id' наибольшего узла
        # Cравниваем Array[id] с его левым и правым дочерними элементами и находим наибольшее значение
        if (left < self.size()) and (self._heap_elem[left] > self._heap_elem[ind]):
            largest = left
        if (right < self.size()) and (self._heap_elem[right] > self._heap_elem[largest]):
            largest = right
        # Если индекс наибольшего и найденного не равны, иначе ничего не делать
        if largest != ind:
            # Поменять местами потомка и родителя
            self._heap_elem[ind], self._heap_elem[largest] = self._heap_elem[largest], self._heap_elem[ind]
            # И вызвать рекурсивно heapify-down для потомка
            self._heapify_down(largest)

    def _heapify_up(self, ind):
        # Ели узел с индексом 'id' и его родитель нарушают свойство кучи
        if ind and (self._heap_elem[self.__get_parent_id(ind)] < self._heap_elem[ind]):
            # Поменять их местами, если свойство кучи нарушено
            self._heap_elem[ind], self._heap_elem[self.__get_parent_id(ind)] = \
                self._heap_elem[self.__get_parent_id(ind)], self._heap_elem[ind]
            # Вызвать heapify-up для родителя
            self._heapify_up(self.__get_parent_id(ind))

    def push(self, key):
        self._heap_elem.append(key)  # Вставить новый элемент в конец вектора
        index = self.size() - 1      # Получить индекс элемента
        self._heapify_up(index)      # Вызов heapify-up от index

    def pop(self):
        try:
            # Если в куче нет элементов, бросить исключение
            if self.size() == 0:
                raise Exception("Heap is empty, stupid man!")
            self._heap_elem[0] = self._heap_elem.pop()        # Заменить корень кучи последним элементом и удалить замененный
            self._heapify_down(0)                             # Вызвать heapify-down для корня
        except Exception as msg:                              # Ловим и печатаем исключение
            print(msg)

    def top(self):
        try:
            # Если в куче нет элементов, бросить исключение
            if self.size() == 0:
                raise Exception("Heap is empty, stupid man!")
            return self._heap_elem[0]
        except Exception as msg:                              # Ловим и печатаем исключение
            print(msg)


class Heap_min:
    def __init__(self):
        self._heap_elem = []

    def __get_parent_id(self, ind):
        return (ind - 1) // 2

    def __get_rchild_id(self, ind):
        return 2 * ind + 2

    def __get_lchild_id(self, ind):
        return 2 * ind + 1

    def size(self):
        return len(self._heap_elem)

    def _heapify_down(self, ind):
        left = self.__get_lchild_id(ind)    # Получить левого потомка узла с индексом 'id'
        right = self.__get_rchild_id(ind)   # Получить правого потомка узла с индексом 'id'
        smallest = ind                      # Запомнить 'id' наименьшего узла
        # Cравниваем Array[id] с его левым и правым дочерними элементами и находим наибольшее значение
        if (left < self.size()) and (self._heap_elem[left] < self._heap_elem[ind]):
            smallest = left
        if (right < self.size()) and (self._heap_elem[right] < self._heap_elem[smallest]):
            smallest = right
        # Если индекс наибольшего и найденного не равны, иначе ничего не делать
        if smallest != ind:
            # Поменять местами потомка и родителя
            self._heap_elem[ind], self._heap_elem[smallest] = self._heap_elem[smallest], self._heap_elem[ind]
            # И вызвать рекурсивно heapify-down для потомка
            self._heapify_down(smallest)

    def _heapify_up(self, ind):
        # Ели узел с индексом 'id' и его родитель нарушают свойство кучи
        if ind and (self._heap_elem[self.__get_parent_id(ind)] > self._heap_elem[ind]):
            # Поменять их местами, если свойство кучи нарушено
            self._heap_elem[ind], self._heap_elem[self.__get_parent_id(ind)] = \
                self._heap_elem[self.__get_parent_id(ind)], self._heap_elem[ind]
            # Вызвать heapify-up для родителя
            self._heapify_up(self.__get_parent_id(ind))

    def push(self, key):
        self._heap_elem.append(key)  # Вставить новый элемент в конец вектора
        index = self.size() - 1      # Получить индекс элемента
        self._heapify_up(index)      # Вызов heapify-up от index

    def pop(self):
        try:
            # Если в куче нет элементов, бросить исключение
            if self.size() == 0:
                raise Exception("Heap is empty, stupid man!")
            self._heap_elem[0] = self._heap_elem.pop()        # Заменить корень кучи последним элементом и удалить замененный
            self._heapify_down(0)                             # Вызвать heapify-down для корня
        except Exception as msg:                              # Ловим и печатаем исключение
            print(msg)

    def top(self):
        try:
            # Если в куче нет элементов, бросить исключение
            if self.size() == 0:
                raise Exception("Heap is empty, stupid man!")
            return self._heap_elem[0]
        except Exception as msg:                              # Ловим и печатаем исключение
            print(msg)


def main():
    iheap = Heap_max()
    dheap = Heap_max()
    imin = Heap_min()
    dmin = Heap_min()

    for i in range(20):
        iheap.push(random.randint(0, 100))
        imin.push(random.randint(0, 100))
        dheap.push(random.randint(0, 100) * 0.28)
        dmin.push(random.randint(0, 100) * 0.28)

    print("Integer max heap: ", end='\t')
    for i in range(20):
        print(iheap.top(), end=' ')
        iheap.pop()
    print("\nInteger min heap: ", end='\t')
    for i in range(20):
        print(imin.top(), end=' ')
        imin.pop()
    print("\nDouble max heap: ", end='\t')
    for i in range(20):
        print(round(dheap.top(), 2), end=' ')
        dheap.pop()
    print("\nDouble min heap: ", end='\t')
    for i in range(20):
        print(round(dmin.top(), 2), end=' ')
        dmin.pop()
    print("\nChecking for emptiness:")
    dheap.top()
    iheap.top()


if __name__ == "__main__":
    main()